src/Microsoft.SqlTools.ServiceLayer/FileBrowser/FileBrowserBase.cs

```diff
@@ -0,0 +1,197 @@
+//
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Globalization;
+using System.Linq;
+using Microsoft.SqlServer.Management.Common;
+using Microsoft.SqlServer.Management.Sdk.Sfc;
+
+namespace Microsoft.SqlTools.ServiceLayer.FileBrowser
+{
+    public struct FileInfo
+    {
+        // Empty for folder
+        public string fileName;
+        public string folderName;
+        public string path;
+        // Includes file name in the path. Empty for folder.
+        public string fullPath;
+    }
+
+    /// <summary>
+    /// Base class for file browser
+    /// </summary>
+    public abstract class FileBrowserBase
+    {
+        private Enumerator enumerator = null;
+        protected object sqlConnection = null;
+
+        protected Enumerator Enumerator
+        {
+            get
+            {
+                return this.enumerator = (this.enumerator ?? new Enumerator());
+            }
+        }
+
+        /// <summary>
+        /// Separator string for components of the file path. Defaults to \ for Windows and / for Linux
+        /// </summary>
+        internal string PathSeparator { get; set; }
+
+        /// <summary>
+        /// Returns the PathSeparator values of the Server.
+        /// </summary>
+        /// <returns>PathSeparator</returns>
+        internal static string GetPathSeparator(Enumerator sfcEnumerator, object sqlConnectionObject)
+        {
+            var req = new Request();
+            req.Urn = "Server";
+            req.Fields = new[] { "PathSeparator" };
+            DataSet ds = sfcEnumerator.Process(sqlConnectionObject, req);
+            string pathSeparator = @"\";
+
+            if (ds.Tables != null && ds.Tables.Count > 0 && ds.Tables[0].Rows != null && ds.Tables[0].Rows.Count > 0)
+            {
+                pathSeparator = Convert.ToString(ds.Tables[0].Rows[0][0], System.Globalization.CultureInfo.InvariantCulture);
+            }
+
+            ds.Dispose();
+            return pathSeparator;
+        }
+
+        /// <summary>
+        /// Enumerates the FileInfo objects associated with drives 
+        /// </summary>
+        /// <param name="enumerator"></param>
+        /// <param name="sqlConnection"></param>
+        /// <returns></returns>
+        internal static IEnumerable<FileInfo> EnumerateDrives(Enumerator enumerator, object sqlConnection)
+        {
+            // if not supplied, server name will be obtained from urn
+            Request req = new Request();
+            int nItems;
+            int i;
+            DataSet ds;
+
+            bool clustered = false;
+
+            req.Urn = "Server/Information";
+            req.Fields = new string[] { "IsClustered", "PathSeparator", "HostPlatform" };
+
+            var pathSeparator = @"\";
+            var hostPlatform = HostPlatformNames.Windows;
+            try
+            {
+                ds = enumerator.Process(sqlConnection, req);
```

Comment:
>enumerator.Process [](start = 21, length = 18)

can this ever return null?  if so you should null check in the various places you assume ds won't be null in this method.

Reply:
it can throw exception if the provided field is not supported on the server. in this case, we're wrapping it in try block because of the "IsClustered" field. if we're using basic field such as "IsFile" we shouldn't need wrap it in try catch.